### PR TITLE
JAMES-2944 Fix missing multipart/related in MIMEMessageConverter

### PR DESCRIPTION
**Pull request by trantienduchn**

### CODE REVIEW

File: server/protocols/jmap-draft/src/main/java/org/apache/james/jmap/draft/methods/MIMEMessageConverter.java [hunk 1]
```diff
@@ -220,6 +216,46 @@ private Multipart createMultipart(CreationMessage newMessage, ImmutableList<Mess
...
+
+    private MultipartBuilder addAttachments(List<MessageAttachment> messageAttachments,
+                                            MultipartBuilder multipartBuilder) {
+        Consumer<MessageAttachment> addAttachment = addAttachment(multipartBuilder);
+        messageAttachments.forEach(addAttachment);
+
+        return multipartBuilder;
+    }
+
+    private Map<Boolean, List<MessageAttachment>> distinguishAttachments(List<MessageAttachment> messageAttachments) {
```
Comment (chibenwa) [inline server/protocols/jmap-draft/src/main/java/org/apache/james/jmap/draft/methods/MIMEMessageConverter.java]: This map with a boolean as key does not express the intent well. Can we find a better way to separate attachments on wether they are  inlined?

Iterating two time the collection seems like a better  option.

File: server/protocols/jmap-draft/src/main/java/org/apache/james/jmap/draft/methods/MIMEMessageConverter.java [hunk 2]
```diff
@@ -220,6 +216,46 @@ private Multipart createMultipart(CreationMessage newMessage, ImmutableList<Mess
         }
     }
 
+    private Multipart createMultipartWithAttachments(CreationMessage newMessage, ImmutableList<MessageAttachment> messageAttachments) throws IOException {
+        MultipartBuilder mixedMultipartBuilder = MultipartBuilder.create(MIXED_SUB_TYPE);
+        Map<Boolean, List<MessageAttachment>> distinguishedAttachments = distinguishAttachments(messageAttachments);
+        List<MessageAttachment> inlineAttachments = Optional.ofNullable(distinguishedAttachments.get(true))
+            .orElse(ImmutableList.of());
+        List<MessageAttachment> normalAttachments = Optional.ofNullable(distinguishedAttachments.get(false))
```
Comment (chibenwa) [inline server/protocols/jmap-draft/src/main/java/org/apache/james/jmap/draft/methods/MIMEMessageConverter.java]: Can we find a more specific world than normal?

File: server/protocols/jmap-draft/src/main/java/org/apache/james/jmap/draft/methods/MIMEMessageConverter.java [hunk 3]
```diff
@@ -220,6 +216,46 @@ private Multipart createMultipart(CreationMessage newMessage, ImmutableList<Mess
...
+            .orElse(ImmutableList.of());
+
+        if (!inlineAttachments.isEmpty()) {
+            MultipartBuilder relatedMultipart = MultipartBuilder.create(RELATED_SUB_TYPE);
+            addBody(newMessage, relatedMultipart);
+            addAttachments(inlineAttachments, relatedMultipart);
```
Comment (chibenwa) [inline server/protocols/jmap-draft/src/main/java/org/apache/james/jmap/draft/methods/MIMEMessageConverter.java]: It seems weird to delegate this to anoyher function.

I would prefer a method returning the mime related multipart...

File: server/protocols/jmap-draft/src/main/java/org/apache/james/jmap/draft/methods/MIMEMessageConverter.java [hunk 4]
```diff
@@ -220,6 +216,46 @@ private Multipart createMultipart(CreationMessage newMessage, ImmutableList<Mess
...
+
+            mixedMultipartBuilder
+                .addBodyPart(Message.Builder.of()
+                    .setBody(relatedMultipart.build())
+                    .build());
+        }
+
+        if (!normalAttachments.isEmpty()) {
+            addBody(newMessage, mixedMultipartBuilder);
```
Comment (chibenwa) [inline server/protocols/jmap-draft/src/main/java/org/apache/james/jmap/draft/methods/MIMEMessageConverter.java]: We dont add the body when we only have inlined attachments?

File: server/protocols/jmap-draft/src/main/java/org/apache/james/jmap/draft/methods/MIMEMessageConverter.java [hunk 5]
```diff
@@ -220,6 +216,46 @@ private Multipart createMultipart(CreationMessage newMessage, ImmutableList<Mess
...
+        MultipartBuilder mixedMultipartBuilder = MultipartBuilder.create(MIXED_SUB_TYPE);
+        Map<Boolean, List<MessageAttachment>> distinguishedAttachments = distinguishAttachments(messageAttachments);
+        List<MessageAttachment> inlineAttachments = Optional.ofNullable(distinguishedAttachments.get(true))
+            .orElse(ImmutableList.of());
+        List<MessageAttachment> normalAttachments = Optional.ofNullable(distinguishedAttachments.get(false))
+            .orElse(ImmutableList.of());
+
+        if (!inlineAttachments.isEmpty()) {
+            MultipartBuilder relatedMultipart = MultipartBuilder.create(RELATED_SUB_TYPE);
+            addBody(newMessage, relatedMultipart);
```
Comment (trantienduchn) [inline server/protocols/jmap-draft/src/main/java/org/apache/james/jmap/draft/methods/MIMEMessageConverter.java]: should not add body twice when having attachments & inline attachments

File: server/protocols/jmap-draft/src/main/java/org/apache/james/jmap/draft/methods/MIMEMessageConverter.java (hunk 2 above)
Comment (trantienduchn) [reply server/protocols/jmap-draft/src/main/java/org/apache/james/jmap/draft/methods/MIMEMessageConverter.java]: IDK how to call it, from what I see on the internet, it's just `attachments`. The best I can do is `inlines` vs `attachments`

Comment (rouazana) [reply server/protocols/jmap-draft/src/main/java/org/apache/james/jmap/draft/methods/MIMEMessageConverter.java]: besideAttachments?

File: server/protocols/jmap-draft/src/main/java/org/apache/james/jmap/draft/methods/MIMEMessageConverter.java (hunk 1 above)
Comment (rouazana) [reply server/protocols/jmap-draft/src/main/java/org/apache/james/jmap/draft/methods/MIMEMessageConverter.java]: or we can return a pair? but maybe that's not really explicit neither

File: server/protocols/jmap-draft/src/main/java/org/apache/james/jmap/draft/methods/MIMEMessageConverter.java (hunk 4 above)
Comment (trantienduchn) [reply server/protocols/jmap-draft/src/main/java/org/apache/james/jmap/draft/methods/MIMEMessageConverter.java]: We add body every time, but the body is wrapped in a multipart/related part when there are inlines

File: server/protocols/jmap-draft/src/main/java/org/apache/james/jmap/draft/methods/MIMEMessageConverter.java (hunk 1 above)
Comment (trantienduchn) [reply server/protocols/jmap-draft/src/main/java/org/apache/james/jmap/draft/methods/MIMEMessageConverter.java]: okay, I took the filter

File: server/protocols/jmap-draft/src/main/java/org/apache/james/jmap/draft/methods/MIMEMessageConverter.java [hunk 6]
```diff
@@ -220,6 +215,46 @@ private Multipart createMultipart(CreationMessage newMessage, ImmutableList<Mess
...
+        MultipartBuilder relatedMultipart = MultipartBuilder.create(RELATED_SUB_TYPE);
+        addBody(newMessage, relatedMultipart);
+
+        return addAttachments(inlines, relatedMultipart)
+            .build();
+    }
+
+    private MultipartBuilder addAttachments(List<MessageAttachment> messageAttachments,
+                                            MultipartBuilder multipartBuilder) {
+        Consumer<MessageAttachment> addAttachment = addAttachment(multipartBuilder);
```
Comment (chibenwa) [inline server/protocols/jmap-draft/src/main/java/org/apache/james/jmap/draft/methods/MIMEMessageConverter.java]: This temporary variable do not add value.

Once you remove it the entire method stop making sens. Can we inline this?

File: server/protocols/jmap-draft/src/main/java/org/apache/james/jmap/draft/methods/MIMEMessageConverter.java [hunk 7]
```diff
@@ -220,6 +215,46 @@ private Multipart createMultipart(CreationMessage newMessage, ImmutableList<Mess
...
+        } else {
+            addBody(newMessage, mixedMultipartBuilder);
+        }
+
+        addAttachments(besideAttachments, mixedMultipartBuilder);
+
+        return mixedMultipartBuilder.build();
+    }
+
+    private Multipart relatedMultipart(CreationMessage newMessage, List<MessageAttachment> inlines) throws IOException {
```
Comment (chibenwa) [inline server/protocols/jmap-draft/src/main/java/org/apache/james/jmap/draft/methods/MIMEMessageConverter.java]: Can't we return the innerMessage here instead, in order to simplify the caller?

File: server/protocols/jmap-draft/src/main/java/org/apache/james/jmap/draft/methods/MIMEMessageConverter.java [hunk 8]
```diff
@@ -220,6 +215,46 @@ private Multipart createMultipart(CreationMessage newMessage, ImmutableList<Mess
...
+    private Multipart createMultipartWithAttachments(CreationMessage newMessage, ImmutableList<MessageAttachment> messageAttachments) throws IOException {
+        MultipartBuilder mixedMultipartBuilder = MultipartBuilder.create(MIXED_SUB_TYPE);
+        List<MessageAttachment> inlineAttachments = messageAttachments.stream()
+            .filter(MessageAttachment::isInline)
+            .collect(Guavate.toImmutableList());
+        List<MessageAttachment> besideAttachments = messageAttachments.stream()
+            .filter(att -> !att.isInline())
+            .collect(Guavate.toImmutableList());
+
+        if (!inlineAttachments.isEmpty()) {
```
Comment (chibenwa) [inline server/protocols/jmap-draft/src/main/java/org/apache/james/jmap/draft/methods/MIMEMessageConverter.java]: Can we turn this condition in a positive form?

File: server/protocols/jmap-draft/src/main/java/org/apache/james/jmap/draft/methods/MIMEMessageConverter.java [hunk 9]
```diff
@@ -220,6 +215,46 @@ private Multipart createMultipart(CreationMessage newMessage, ImmutableList<Mess
         }
     }
 
+    private Multipart createMultipartWithAttachments(CreationMessage newMessage, ImmutableList<MessageAttachment> messageAttachments) throws IOException {
+        MultipartBuilder mixedMultipartBuilder = MultipartBuilder.create(MIXED_SUB_TYPE);
+        List<MessageAttachment> inlineAttachments = messageAttachments.stream()
+            .filter(MessageAttachment::isInline)
+            .collect(Guavate.toImmutableList());
+        List<MessageAttachment> besideAttachments = messageAttachments.stream()
+            .filter(att -> !att.isInline())
```
Comment (chibenwa) [inline server/protocols/jmap-draft/src/main/java/org/apache/james/jmap/draft/methods/MIMEMessageConverter.java]: att?

File: server/protocols/jmap-draft/src/test/java/org/apache/james/jmap/draft/methods/MIMEMessageConverterTest.java [hunk 10]
```diff
@@ -571,23 +578,26 @@ public void convertToMimeShouldSetEmptyTextBodyWhenProvided() {
...
-    @Test
-    public void convertToMimeShouldAddAttachmentWhenOne() {
-        // Given
-        MIMEMessageConverter sut = new MIMEMessageConverter();
+    @Nested
+    class WithAttachments {
 
-        CreationMessage testMessage = CreationMessage.builder()
+        @Test
+        void convertToMimeShouldAddAttachmentWhenOne() {
```
Comment (chibenwa) [inline server/protocols/jmap-draft/src/test/java/org/apache/james/jmap/draft/methods/MIMEMessageConverterTest.java]: one what?

**Comment by Arsnael** [comment]
CI restart...
test this please

File: server/protocols/jmap-draft/src/main/java/org/apache/james/jmap/draft/methods/MIMEMessageConverter.java [hunk 11]
```diff
@@ -220,6 +215,43 @@ private Multipart createMultipart(CreationMessage newMessage, ImmutableList<Mess
...
+    private Multipart createMultipartWithAttachments(CreationMessage newMessage, ImmutableList<MessageAttachment> messageAttachments) throws IOException {
+        MultipartBuilder mixedMultipartBuilder = MultipartBuilder.create(MIXED_SUB_TYPE);
+        List<MessageAttachment> inlineAttachments = messageAttachments.stream()
+            .filter(MessageAttachment::isInline)
+            .collect(Guavate.toImmutableList());
+        List<MessageAttachment> besideAttachments = messageAttachments.stream()
+            .filter(attachment -> !attachment.isInline())
+            .collect(Guavate.toImmutableList());
+
+        if (inlineAttachments.size() > 0) {
```
Comment (chibenwa) [inline server/protocols/jmap-draft/src/main/java/org/apache/james/jmap/draft/methods/MIMEMessageConverter.java]: `.isEmpty()` is more straight forward

File: server/protocols/jmap-draft/src/main/java/org/apache/james/jmap/draft/methods/MIMEMessageConverter.java [hunk 12]
```diff
@@ -220,6 +215,43 @@ private Multipart createMultipart(CreationMessage newMessage, ImmutableList<Mess
...
+
+        return Message.Builder.of()
+            .setBody(addAttachments(inlines, relatedMultipart)
+                .build())
+            .build();
+    }
+
+    private MultipartBuilder addAttachments(List<MessageAttachment> messageAttachments,
+                                            MultipartBuilder multipartBuilder) {
+        messageAttachments.forEach(addAttachment(multipartBuilder));
```
Comment (chibenwa) [inline server/protocols/jmap-draft/src/main/java/org/apache/james/jmap/draft/methods/MIMEMessageConverter.java]: Really I believe we can inline this...

**Comment by trantienduchn** [comment]
approved

**Comment by chibenwa** [comment]
Was :green_apple:

**Comment by trantienduchn** [comment]
was :green_apple: got enough approvals

**Comment by chibenwa** [comment]
Merged

**Pull request closed without merging**